src/components/VmsList/VmCardList.js

```diff
@@ -0,0 +1,102 @@
+import React, { useEffect } from 'react'
+import PropTypes from 'prop-types'
+import { connect } from 'react-redux'
+import { msg } from '_/intl'
+import { getByPage } from '_/actions'
+import { filterVms, sortFunction } from '_/utils'
+
+import useInfiniteScroll from '@closeio/use-infinite-scroll'
+import Vm from './Vm'
+import Pool from './Pool'
+
+import style from './style.css'
+
+/**
+ * Use Patternfly 'Single Select Card View' pattern to show every VM and Pool
+ * available to the current user.
+ */
+const VmCardList = ({ vms, alwaysShowPoolCard, fetchMoreVmsAndPools }) => {
+  const sort = vms.get('sort').toJS()
+  const filters = vms.get('filters').toJS()
+
+  // Filter the VMs (1. apply the filter bar criteria, 2. only show Pool VMs if the Pool exists)
+  const filteredVms = vms.get('vms')
+    .filter(vm => filterVms(vm, filters))
+    .filter(vm => vm.getIn(['pool', 'id'], false) ? !!vms.getIn(['pools', vm.getIn(['pool', 'id'])], false) : true)
+    .toList()
+    .map(vm => vm.set('isVm', true))
+
+  // Filter the Pools (only show a Pool card if the user can currently 'Take' a VM from it)
+  const filteredPools = vms.get('pools')
+    .filter(pool => alwaysShowPoolCard || (pool.get('vmsCount') < pool.get('maxUserVms') && pool.get('size') > 0 && filterVms(pool, filters)))
+    .toList()
+
+  // Display the VMs and Pools together, sorted nicely
+  const vmsAndPools = [ ...filteredVms, ...filteredPools ].sort(sortFunction(sort))
+
+  // Handle the infinite scroll and pagination
+  const hasMore = vms.get('vmsExpectMorePages') || vms.get('poolsExpectMorePages')
+  const [ page, sentinelRef, scrollerRef ] = useInfiniteScroll({ hasMore, distance: 0 })
+
+  useEffect(() => { // `refreshListPage` handles loading the first page of data
+    if (page > 0) {
+      fetchMoreVmsAndPools()
+    }
+  }, [ page ])
+
+  useEffect(() => {
+    if (!scrollerRef.current || !sentinelRef.current) {
+      return
+    }
+
+    //
+    // If a page fetch doesn't pull enough entities to push the sentinel out of view
+    // underlying IntersectionObserver doesn't fire another event, and the scroller
+    // gets stuck.  Manually check if the sentinel is in view, and if it is, fetch
+    // more data.  The effect is only run when the `vms` part of the redux store is
+    // updated.
+    //
+    const scrollRect = scrollerRef.current.getBoundingClientRect()
+    const scrollVisibleTop = scrollRect.y
+    const scrollVisibleBottom = scrollRect.y + scrollRect.height
+
+    const sentinelRect = sentinelRef.current.getBoundingClientRect()
+    const sentinelTop = sentinelRect.y
+    const sentinelBottom = sentinelRect.y + sentinelRect.height
+
+    const sentinelStillInView = sentinelBottom >= scrollVisibleTop && sentinelTop <= scrollVisibleBottom
+    if (sentinelStillInView) {
+      fetchMoreVmsAndPools()
```

Comment:
It seems that we might end up loading all pages one-after-another.
This may happen if filtering is used i.e. 
1. filter removes all vm's from newly loaded page so sentinel was not moved. 
2. vms has changed so hook is triggered and fetches new page
3. new page is also filtered out so again sentinel was not moved  ....

Perhaps filtering can be moved to saga-layer and merged with paging logic? It could fix some corner cases

Reply:
Filtering on the front end, paging on the back end and infinite scrolling UX don't work so well together.  

For a reference interaction set: your page size is set to 4, and you have 26 VMs named Alpha to Zulu, and you see 4x3 VMs on the screen.   A total of 3 pages are fetched up front to fill the screen (leaving 4 unfetched pages).  Now the user types in a filter of "ul".  Some cards disappear and the sentinel moves up, triggering more page fetches.  Eventually 12 cards are in view and the extra fetching stops.  If there are still pages remaining (I didn't check to see how many phonetic alphabet words have "ul" in them), then they're accessible on user scroll.  There could be a weird no-op scroll delay if a new page is fetched and no new cards match the filter.

So maybe there isn't a problem?

I'm trying to reason out when it could go bad, but really I think more pages need to be fetched to be able to fill up the screen with visible things.  Worst part will be when the user has to scroll and wait for multiple pages to be fetched until enough visible cards exist or they're hit the end of the list.

I'm just trying to think of the sentinel div coming into view as the same thing as a user clicking "load more".

Reply:
I was able to test the scenario described above and actually loading-until-full feels the right approach. 
The part that could be improved is triggering the effect on every change of `vms` object. This includes property changes.
Fetch should be probably triggered only if IDs change since only this will trigger reloading cards.

Reply:
I'm not quite sure what your concern is here.

Paged fetching runs ok and new pages of VM/Pools are pushed into redux state.  The new data causes react to render everything again, which is required.  On a refresh cycle, the new data still needs to be evaluated and differences rendered.  Something like a VM starting and now the status icon and action buttons need to be rendered differently.  All of that works as appropriate.